### DISCUSSION TITLE
mitmproxy stops processing flows with large amount of traffic at once

### DISCUSSION THREAD

**Discussion by malwareinfosec** [mitmproxy/mitmproxy#6901] [General]
I'm running a performance test with mitmproxy where I generate a lot of web traffic in a short period of time (i.e. opening 20 different URLs at once and then repeating 2 or more times).
Very shortly after, mitmproxy stops processing new flows. It has not crashed and is not frozen, there are just no more flows being processed. The system also reports an issue with the proxy. I can quit mitmproxy using the 'q' command and it lets me quit the application without issue.
I've tried the same test with mitmdump (thinking this could be a memory problem) with the same outcome.

If I relaunch mitmproxy after, it starts working as expected, that is until I submit a large amount of traffic at once again. Any idea what is causing this?

![image](https://github.com/mitmproxy/mitmproxy/assets/25351665/ce4d53f8-0a01-41a4-b790-24f319e26793)

**Comment by Prinzhorn** [comment]
Before looking into this, can you try with the hottest latest build? Download: https://github.com/mitmproxy/mitmproxy/actions/runs/9439274286/artifacts/1583458000 (via https://github.com/mitmproxy/mitmproxy/actions/runs/9439274286)

**Reply by malwareinfosec** [reply]
Thanks, I tried the newest binary and was able to reproduce.

I ran a loop of:
- open 20 browser tabs
- pause for 30 seconds
- kill browser
- repeat

After about 10 minutes, the proxy was no longer working:

![image](https://github.com/mitmproxy/mitmproxy/assets/25351665/c3e7b9db-0a5f-4ad6-ade3-18795759e10a)

**Reply by mhils** [reply]
Does this repro on Linux? If yes, can you send SIGUSR1/SIGUSR2 to the mitmproxy process when it is stuck and post the output here?

**Reply by malwareinfosec** [reply]
It did not repro on Linux! This only happened on Windows (Win10 VM, and Win11 bare metal).

**Reply by mhils** [reply]
Thanks. That makes it tricky to debug. You could maybe modify https://github.com/mitmproxy/mitmproxy/blob/v10.3.0/mitmproxy/utils/debug.py#L128 to print debug info after a fixed amount of time instead of relying on signals?
